core/src/main/scala/org/http4s/HeaderV2.scala

```diff
@@ -0,0 +1,292 @@
+    * multiple values.
+    */
+  def parse(headerValue: String): Option[A]
+}
+object Header {
+  case class Raw(name: CIString, value: String, recurring: Boolean = true)
+  object Raw {
+    implicit val showForRawHeaders: Show[Raw] =
+      h => s"${h.name.show}: ${h.value}"
+
+    implicit lazy val orderForRawHeaders: Order[Raw] =
+      Order.whenEqual(
+        Order.by(_.name),
+        Order.by(_.value)
+      )
+  }
+
+  /** Classifies custom headers into `Single` headers, which can only
+    * appear once, and `Recurring` headers, which can appear multiple
+    * times.
+    */
+  sealed trait Type
+  case class Single() extends Type
+  case class Recurring() extends Type
+
+  def apply[A](implicit ev: Header[A, _]): ev.type = ev
+
+  /** Target for implicit conversions to Header.Raw from custom
+    * headers and key-value pairs.
+    *
+    * A method taking variadic `ToRaw` arguments will allow taking
+    * heteregenous arguments, provided they are either:
+    *
+    * - A value of type `A`  which has a `Header[A]` in scope
+    * - A (name, value) pair of `String`, which is treated as a `Recurring` header
+    * - A `Header.Raw`
+    *
+    * @see [[org.http4s.Headers$.apply]]
+    */
+  sealed trait ToRaw {
+    def value: Header.Raw
+  }
+  object ToRaw {
+    implicit def rawToRaw(h: Header.Raw): Header.ToRaw = new Header.ToRaw {
+      val value = h
+    }
+
+    implicit def keyValuesToRaw(kv: (String, String)): Header.ToRaw = new Header.ToRaw {
+      val value = Header.Raw(CIString(kv._1), kv._2)
+    }
+
+    implicit def customHeadersToRaw[H](h: H)(implicit H: Header[H, _]): Header.ToRaw =
+      new Header.ToRaw {
+        val value = Header.Raw(H.name, H.value(h))
+      }
+  }
+
+  /** Abstracts over Single and Recurring Headers
+    */
+  sealed trait Select[A] {
+    type F[_]
+
+    /** Transform this header into a [[Header.Raw]]
+      */
+    def toRaw(a: A): Header.Raw
+
+    /** Selects this header from a list of [[Header.Raw]]
+      */
+    def from(headers: List[Header.Raw]): Option[F[A]]
+  }
+  object Select {
+    def fromRaw[A](h: Header.Raw)(implicit ev: Header[A, _]): Option[A] =
+      (h.name == Header[A].name).guard[Option] >> Header[A].parse(h.value)
+
+    implicit def singleHeaders[A](implicit
+        h: Header[A, Header.Single]): Select[A] { type F[B] = Id[B] } =
+      new Select[A] {
+        type F[B] = Id[B]
+
+        def toRaw(a: A): Header.Raw =
+          Header.Raw(h.name, h.value(a), recurring = false)
+
+        def from(headers: List[Header.Raw]): Option[A] =
+          headers.collectFirst(Function.unlift(fromRaw(_)))
+      }
+
+    implicit def recurringHeaders[A](implicit
+        h: Header[A, Header.Recurring]): Select[A] { type F[B] = NonEmptyList[B] } =
+      new Select[A] {
+        type F[B] = NonEmptyList[B]
+
+        def toRaw(a: A): Header.Raw =
+          Header.Raw(h.name, h.value(a))
+
+        def from(headers: List[Header.Raw]): Option[NonEmptyList[A]] =
```

Comment:
So we're choosing not to combine them then, but to return a List of them.  We might have `Some(NonEmptyList(Accept(NonEmptyList(...))))`?

Reply:
It's hard to combine them, because we no longer have the information about the fact that there is ` values` field of type `NonEmptyList` inside our `A`.  We could introduce the semigroup idea you had, this is mostly to get the mechanics for `get` working, rather than a firm decision on how to combine them (or whether to)

Reply:
Ok we reached a solution, I'll implement it tomorrow: if you provide an instance for `Semigroup[A]` and `Header[A, Recurring]`, we merge them, if you don't, you get a NonEmptyList